src/restic_fmt_plug.c

```diff
@@ -0,0 +1,334 @@
+#include "formats.h"
+#include "jumbo.h"
+#include "misc.h"
+#include "options.h"
+#include "params.h"
+#include "poly1305-donna/poly1305-donna.h"
+#include "yescrypt/yescrypt.h"
+
+#define FORMAT_NAME "Restic Repository"
+#define FORMAT_LABEL "restic"
+#define FORMAT_TAG "$restic$"
+#define TAG_LENGTH sizeof(FORMAT_TAG) - 1
+#define ALGORITHM_NAME "PBKDF2/scrypt Poly1305"
+#define PLAINTEXT_LENGTH 125
+#define BINARY_SIZE 16 /* MAC */
+#define BINARY_ALIGN 4
+#define DATA_SIZE 160
+#define SALT_ALIGN 4
+#define MAC_SIZE 16
+#define NONCE_SIZE 16
+#define BENCHMARK_COMMENT ""
+#define BENCHMARK_LENGTH 7
+#define MIN_KEYS_PER_CRYPT 1
+#define MAX_KEYS_PER_CRYPT 1
+
+static int max_threads;
+static yescrypt_local_t *local;
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static uint32_t (*crypt_out)[BINARY_SIZE / sizeof(uint32_t)];
+
+typedef struct restic_salt {
+	int N;
+	int r;
+	int p;
+	char salt[64];
+	unsigned char data[DATA_SIZE];
+} restic_salt;
+
+static restic_salt *cur_salt;
+
+static void init(struct fmt_main *self)
+{
+	max_threads = 1;
+	local = mem_alloc(sizeof(*local) * max_threads);
+	int i;
+	for (i = 0; i < max_threads; i++)
+		yescrypt_init_local(&local[i]);
+
+	saved_key = mem_calloc(self->params.max_keys_per_crypt, sizeof(*saved_key));
+	crypt_out = mem_calloc(self->params.max_keys_per_crypt, sizeof(*crypt_out));
+}
+
+static void done(void)
+{
+	int i;
+	for (i = 0; i < max_threads; i++)
+		yescrypt_free_local(&local[i]);
+	MEM_FREE(local);
+	MEM_FREE(saved_key);
+	MEM_FREE(crypt_out);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *ctcopy, *keeptr, *p;
+
+	if (strncmp(ciphertext, FORMAT_TAG, TAG_LENGTH) != 0)
+		return 0;
+
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += TAG_LENGTH;
+	if (((p = strtokm(ctcopy, "*")) == NULL) || (strcmp("scrypt", p) != 0)) /* KDF */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !isdec(p)) /* scrypt: N */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !isdec(p)) /* scrypt: r */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !isdec(p)) /* scrypt: p */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !ishexn(p, 64)) /* scrypt: salt */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !ishexn(p, 160)) /* restic: data */
+		goto err;
+
+	MEM_FREE(keeptr);
+	return 1;
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+void *restic_get_binary(char *ciphertext)
+{
+	static unsigned char c[BINARY_SIZE];
+	char *p;
+	int i;
+	p = strrchr(ciphertext, '*') + 1;
+	// Restic design doc: computed MAC will not match the last 16 bytes of the data
+	p += (160 - 16) * 2;
+	for (i = 0; i < BINARY_SIZE; i++) {
+		c[i] = (atoi16[ARCH_INDEX(*p)] << 4) | atoi16[ARCH_INDEX(p[1])];
+		p += 2;
+	}
+	return c;
+}
+
+void *restic_get_salt(char *ciphertext)
+{
+	char *ctcopy = strdup(ciphertext);
+	char *keeptr = ctcopy;
+	int i;
+	char *p;
+	restic_salt *cur_salt;
+
+	cur_salt = mem_calloc_tiny(sizeof(restic_salt), MEM_ALIGN_SIMD);
```

Comment:
Why `MEM_ALIGN_SIMD`? I think this should be `SALT_ALIGN`.

Reply:
Indeed, that was also just a boilerplate copy.